astacus/coordinator/list.py

```diff
@@ -3,19 +3,37 @@
 See LICENSE for details
 
 """
-
 from astacus.common import ipc, magic
-from astacus.common.storage import MultiStorage
-
-
-def _iter_backups(storage):
+from astacus.common.storage import JsonStorage, MultiStorage
+from collections.abc import Iterator
+
+
+def compute_unique_snapshot_file_stats(manifest: ipc.BackupManifest) -> tuple[int, int]:
+    """Compute stats over unique files as determined by their hex digest."""
+    seen_hexdigests = set()
+    num_files = 0
+    total_size = 0
+    for snapshot_result in manifest.snapshot_results:
+        if not snapshot_result.state:
+            num_files += snapshot_result.files
+            total_size += snapshot_result.total_size
+        else:
+            for snapshot_file in snapshot_result.state.files:
+                if snapshot_file.hexdigest in seen_hexdigests:
+                    continue
+                seen_hexdigests.add(snapshot_file.hexdigest)
+                num_files += 1
+                total_size += snapshot_file.file_size
```

Comment:
```suggestion
                num_files += 1
                total_size += snapshot_file.file_size
            for snapshot_file in snapshot_result.state.files:
                seen_hexdigests.add(snapshot_file.hexdigest)
```

I think we shouldn't de-duplicate within a single `snapshot_result`, while still de-duplicating between different `snapshot_result`:

We don't want to count twice the data that is identical between two different nodes because they are in the same shard (or replicas, depending on the service type of the backup).

But we also don't want the original uncompressed total size to be smaller than the volume on data on a single node, and if someone has created 10 identical 1GB tables, which are taking 10GB on the disk, we want the original size to reflect that (this will matter when estimating whether we can restore the backup on a specific disk of fixed size).

(that means the function name should also change).

Reply:
Thank you for the comment. The function was reworked to aggregate size and file count information globally on a per-hexdigest basis and apply a reduction to the stats gathered.